sphinx/util/inspect.py

```diff
@@ -315,6 +315,114 @@ def is_builtin_class_method(obj: Any, attr_name: str) -> bool:
+        raise TypeError("can't compute signature for built-in type {}".format(subject))
+
+    try:
+        signature = inspect.signature(subject)
+        parameters = list(signature.parameters.values())
+        return_annotation = signature.return_annotation
+    except IndexError:
+        # Until python 3.6.4, cpython has been crashed on inspection for
+        # partialmethods not having any arguments.
+        # https://bugs.python.org/issue33009
+        if hasattr(subject, '_partialmethod'):
+            parameters = []
+            return_annotation = inspect.Parameter.empty
+        else:
+            raise
+
+    try:
+        # Update unresolved annotations using ``get_type_hints()``.
+        annotations = typing.get_type_hints(subject)
+        for i, param in enumerate(parameters):
+            if isinstance(param.annotation, str) and param.name in annotations:
+                parameters[i] = param.replace(annotation=annotations[param.name])
+        if 'return' in annotations:
+            return_annotation = annotations['return']
+    except Exception:
+        # ``get_type_hints()`` does not support some kind of objects like partial,
+        # ForwardRef and so on.
+        pass
+
+    if bound_method:
+        if inspect.ismethod(subject):
+            # ``inspect.signature()`` considers the subject is a bound method and removes
+            # first argument from signature.  Therefore no skips are needed here.
+            pass
+        else:
+            if len(parameters) > 0:
+                parameters.pop(0)
+
+    return inspect.Signature(parameters, return_annotation=return_annotation)
+
+
+def stringify_signature(sig: inspect.Signature, show_annotation: bool = True,
+                        show_return_annotation: bool = True) -> str:
+    """Stringify a Signature object.
+
+    :param show_annotation: Show annotation in result
+    """
+    args = []
+    last_kind = None
+    for param in sig.parameters.values():
+        # skip first argument if subject is bound method
+        arg = StringIO()
+
+        # insert '*' between POSITIONAL args and KEYWORD_ONLY args::
+        #     func(a, b, *, c, d):
+        if param.kind == param.KEYWORD_ONLY and last_kind in (param.POSITIONAL_OR_KEYWORD,
+                                                              param.POSITIONAL_ONLY,
+                                                              None):
+            args.append('*')
+
+        if param.kind in (param.POSITIONAL_ONLY,
+                          param.POSITIONAL_OR_KEYWORD,
+                          param.KEYWORD_ONLY):
+            arg.write(param.name)
+            if show_annotation and param.annotation is not param.empty:
+                arg.write(': ')
+                arg.write(stringify_annotation(param.annotation))
+            if param.default is not param.empty:
+                if param.annotation is param.empty or show_annotation is False:
```

Comment:
It would be better to invert this condition. Because other conditions are inverted.